chore(TS-52): swagger 설정

stableh opened the pull request:
## 📋 이슈 번호
- [TS-52](https://jeez.atlassian.net/browse/TS-52?atlOrigin=eyJpIjoiZjI0NTJkNDhjMWE4NGQ1ZDkxMzIyYTAzNGU5YjMyYTUiLCJwIjoiaiJ9)

## ✅ 변경 사항
- 스프링 시큐리티의 httpBasic을 이용하여 로그인을 하여 스웨거에 접근하도록 설정하였습니다.
- Swagger가 아닌 SpringDoc을 이용하여 작성하였습니다.

[TS-52]: https://jeez.atlassian.net/browse/TS-52?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

stableh committed:
chore: swagger 초기 세팅

stableh committed:
chore: submodule 업데이트

stableh committed:
chore: swagger 로그인 설정

stableh committed:
chore: Auth 도메인 swagger 설정

stableh committed:
chore: registraion 도메인 swagger 설정

stableh committed:
chore: schedule 도메인 swagger 설정

stableh committed:
chore: wishlist 도메인 swagger 설정

zhy2on commented on src/main/java/com/tutorialsejong/courseregistration/domain/registration/swagger/RegisterCourseOperation.java:
```diff
@@ -1,4 +1,31 @@
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.METHOD})
```
swagger 설정 하나하나 다 꼼꼼히 하시느라 너무 수고하셨습니다ㅠㅠ
1. java annotation
`@Target`으로 어노테이션이 적용될 수 있는 대상을 제한해주고
`@Retention`으로 어노테이션 생명주기를 정해주고

2. swagger annotation
`@Operation`으로는 해당 api에 대한 상세 정보를,
`@ApiResponses`로는 가능한 응답들을 적어주는 거네요

여기서 Retention을 왜 Runtime으로 하는지 궁금했는데 Swagger가 런타임에 어노테이션 정보를 읽어서 API 문서를 생성해야 하기 때문이었네요!

Anhye0n replied:
감사합니다!
원래는 기존 Controller에 스웨거 어노테이션으로 추가하는 방식으로 작성했습니다만, 그렇게 하니 코드 가독성이 너무 안좋아져 새로 코드를 추가하여 작성하였습니다.

zhy2on reviewed (changes requested):
수고 많으셨습니다😭
파일 끝 개행 넣기, 정렬 맞추기는 인텔리제이 설정 통일하면 좋을 거 같아요!
SecurityFilterChain 관련 부분도 검토 한 번 부탁드립니다..!

Anhye0n commented on src/main/java/com/tutorialsejong/courseregistration/common/security/AccessDeniedFilter.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    // 예시 응답 DTO
+    static class ErrorResponse {
```
기존에 ErrorResponse를 까먹고 새로 작성했었네요! 수정하겠습니다.

Anhye0n commented on src/main/java/com/tutorialsejong/courseregistration/common/security/JwtAuthenticationFilter.java:
```diff
@@ -57,6 +57,6 @@ private void enhanceAuthenticationWithRequestDetails(Authentication authenticati
     protected boolean shouldNotFilter(HttpServletRequest request) {
         String path = request.getRequestURI();
-        return path.equals("/api/auth/login") || path.equals("/api/auth/refresh") || path.equals("/api/auth/withdraw");
+        return path.equals("/api/auth/login") || path.equals("/api/auth/refresh") || path.equals("/api/auth/withdraw") || path.equals("/docs/swagger-ui/**") || path.equals("/v3/api-docs/**");
```
새로 추가하고, 기존에 작성한 코드를 지우는걸 깜빡했었네요.. 제거했습니다!

Anhye0n commented on src/main/java/com/tutorialsejong/courseregistration/domain/registration/service/CourseRegistrationService.java:
```diff
@@ -114,10 +115,10 @@ public CourseRegistrationResponse registerCourse(String studentId, Long schedule
 
     @Transactional(readOnly = true)
-    public List<Schedule> getRegisteredCourses(String studentId) {
+    public List<CourseRegistrationScheduleResponse> getRegisteredCourses(String studentId) {
```
👍

Anhye0n commented on src/main/java/com/tutorialsejong/courseregistration/common/config/SecurityConfig.java:
```diff
@@ -32,15 +37,56 @@ public class SecurityConfig {
                 .headers(headers -> headers
-                        .contentSecurityPolicy(csp -> csp
-                                .policyDirectives("frame-ancestors 'self'"))
+                        .contentSecurityPolicy(csp -> csp.policyDirectives("frame-ancestors 'self'"))
```
맥북을 초기화하며 기존의 정렬 세팅이 날아간 것 같습니다. 다시 세팅하도록 하겠습니다.

stableh committed:
refactor: shouldNotFilter 메서드 수정

stableh committed:
refactor: 에러 응답 일관성 수정

Anhye0n commented on src/main/java/com/tutorialsejong/courseregistration/common/security/AccessDeniedFilter.java:
덕분에 저도 이 구조에 대해 공부를 하게 된 것 같습니다.
1. 사용자가 웹페이지에 접근
2. 401 응답 및 WWW-Authenticate 헤더 확인 시, 브라우저에서 로그인 alert 띄움
3. 401 응답 = AuthenticationEntryPoint이기 때문에 SwaggerAuthenticationEntryPoint에 헤더와 401 응답을 설정해주면 정상적으로 예외처리 진행

```java
public class SwaggerAuthenticationEntryPoint implements AuthenticationEntryPoint {
    @Override
    public void commence(HttpServletRequest request, HttpServletResponse response,
                         AuthenticationException authException) throws IOException, ServletException {
        response.addHeader("WWW-Authenticate", "Basic realm=\"Tutorial Sejong Swagger\"");
        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);

        ErrorResponse.from(GlobalErrorCode.UNAUTHORIZED)
                .writeTo(response);
    }
}
```

위 과정으로 작성했습니다. 403 에러는 Handler로, 401 에러는 EntryPointer로 처리했기 때문에 기존에 Filter는 필요가 없어져 제거하였습니다.

stableh committed:
refactor: Swagger SecurityConfig 구조 수정

zhy2on committed:
refactor: Swagger 관련 컴포넌트 개선

- SwaggerAccessDeniedHandler를 common/security 패키지로 이동
- SwaggerAuthenticationEntryPoint에서 불필요한 setStatus 제거 (ErrorResponse에서 처리)
- 불필요한 ServletException throws 제거

zhy2on committed:
chore: 파일 끝 개행 삭제

zhy2on committed:
chore: redundant parameter 제거

- required=false 제거
- 스프링독(SpringDoc) OpenAPI 문서의 @Parameter 어노테이션에서 required 속성의 기본값이 이미 false이므로 불필요.

zhy2on committed:
chore: 파일 끝 개행 추가
